ECS architecture docs for Beta (#11043)
hashicorp/consul: website/content/docs/ecs/architecture.mdx

- **T1:** The `sidecar-proxy` container starts. It runs Envoy by executing `envoy -c <path-to-bootstrap-json>`.
- **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, your application containers are started since all Consul machinery is ready to service requests. The only running containers are `consul-client`, `sidecar-proxy`, and your application container(s).

### Automatic ACL Token Provisioning

Consul ACL tokens secure communication between agents and services.
The following containers in a task require an ACL token:

- `consul-client`: The Consul client uses a token to authorize itself with Consul servers.
  All `consul-client` containers share the same token.
- `mesh-init`: The `mesh-init` container uses a token to register the service with Consul.
  This token is unique for the Consul service, and is shared by instances of the service.

The ACL controller automatically creates ACL tokens for mesh-enabled tasks in an ECS cluster.

> The `acl-controller` Terraform module creates the ACL token used by `consul-client` containers, and
> then starts the ACL controller task. The controller watches for tasks in the cluster. It checks tags

Reviewer: Rephrasing a bit to clarify that the controller task will create the token rather than the module.
```suggestion
The `acl-controller` Terraform module the ACL controller task. The controller creates the ACL token used by `consul-client` containers at startup and then watches for tasks in the cluster. It checks tags
```

![Consul on ECS Architecture](/img/consul-ecs-arch.png)

> 

Reviewer 2: It might align better with reader expectations to move the description of the diagram above it, e.g.:
"The following diagram shows the main components of the Consul architecture when deployed to an ECS cluster:"

I would keep the bulleted list describing the components below the graphic. 

Just from a reader psychology perspective, this would set up a "here's what to expect", "here's the thing", "here's more info about the thing" format.

As shown above, these are the main components to the architecture for a secure installation:

1. **Consul Servers:** Production-ready Consul server cluster
1. **Application tasks:** Runs user application containers along with two helper containers:
   1. **Consul Client:** The Consul client container runs Consul. The Consul client communicates
      with the Consul server and configures the Envoy proxy sidecar. This communication
      is called _control plane_ communication.
   1. **Sidecar Proxy:** The sidecar proxy container runs [Envoy](https://envoyproxy.io/). All requests
      to and from the application container(s) run through the sidecar proxy. This communication
      is called _data plane_ communication.
1. **ACL Controller:** Automatically provisions Consul ACL tokens for Consul clients and service mesh services
   in an ECS Cluster.

For more information about how Consul works in general, see Consul's [Architecture Overview](/docs/architecture).

> In addition to the long-running Consul Client and Sidecar Proxy containers, the `mesh-init` container runs

Reviewer 2: ```suggestion
In addition to the long-running Consul client and sidecar proxy containers, the `mesh-init` container runs
```
I think this is more consistent with our style guide--at least, other references use lower case. But ignore if I'm mistaken.

at startup and sets up initial configuration for Consul and Envoy.

### Task Startup

This diagram shows the timeline of a task starting up and all its containers:

![Task Startup Timeline](/img/ecs-task-startup.png)

- **T0:** ECS starts the task. The `consul-client` and `mesh-init` containers start:
  - `consul-client` uses the `retry-join` option to join the Consul cluster
  - `mesh-init` registers the service for this task and its sidecar proxy into Consul. It runs `consul connect envoy -bootstrap` to generate Envoy’s bootstrap JSON file and write it to a shared volume. After registration and bootstrapping, `mesh-init` exits.
- **T1:** The `sidecar-proxy` container starts. It runs Envoy by executing `envoy -c <path-to-bootstrap-json>`.

> - **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, the user’s application containers are started since all the Consul machinery is ready to service requests. The only running containers are `consul-client`, `sidecar-proxy` and the user’s application container(s).

Reviewer 2: ```suggestion
- **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, your application containers are started since all Consul machinery is ready to service requests. The only running containers are `consul-client`, `sidecar-proxy`, and your application container(s).
```
I believe the style guide wants us to address the reader directly and avoid 'the user'. It helps maintain an active voice. But this guidance only applies to the Consul practitioner, so please ignore this suggestion if we are referring to an end user other than the Consul user.

Author: Ah. Hmm. I'm not too familiar actually. It is an architecture description, rather than guide telling a reader what to do if that matters.

I do think the second person there helps clarify what we set up for you, versus what you need to bring (application containers)

In addition to the long-running Consul Client and Sidecar Proxy containers, the `mesh-init` container runs
at startup and sets up initial configuration for Consul and Envoy.

### Task Startup

This diagram shows the timeline of a task starting up and all its containers:

![Task Startup Timeline](/img/ecs-task-startup.png)

- **T0:** ECS starts the task. The `consul-client` and `mesh-init` containers start:
  - `consul-client` uses the `retry-join` option to join the Consul cluster
  - `mesh-init` registers the service for this task and its sidecar proxy into Consul. It runs `consul connect envoy -bootstrap` to generate Envoy’s bootstrap JSON file and write it to a shared volume. After registration and bootstrapping, `mesh-init` exits.
- **T1:** The `sidecar-proxy` container starts. It runs Envoy by executing `envoy -c <path-to-bootstrap-json>`.
- **T2:** The `sidecar-proxy` container is marked as healthy by ECS. It uses a health check that detects if its public listener port is open. At this time, the user’s application containers are started since all the Consul machinery is ready to service requests.

> - **T3:** Consul marks the service as healthy by running the health checks specified in the task Terraform. The service will now receive traffic. At this time the only running containers are `consul-client`, `sidecar-proxy` and the user’s application container(s).

Reviewer 3: do we support heath checks in the task tf?

Author: Nope, no health Consul native health checks in Terraform. Fixed! On that point, we can then remove T3 then as well?

Reviewer 3: Yep.
